controller/src/main/java/io/pravega/controller/server/bucket/BucketManagerLeader.java

```diff
@@ -0,0 +1,210 @@
+    private final AtomicBoolean suspended = new AtomicBoolean(false);
+    // The minimum interval between any two redistribution operations. The minimum duration is not guaranteed when leadership
+    // moves across controllers. Since this is uncommon and there are no significant side-effects to it, we don't
+    // handle this scenario.
+    private final Duration minBucketRedistributionIntervalInSeconds;
+
+    // The controller to bucket distributor.
+    @Getter
+    private final BucketDistributor bucketDistributor;
+    // Service type
+    private final BucketStore.ServiceType serviceType;
+    private final ScheduledExecutorService executorService;
+
+    public BucketManagerLeader(BucketStore bucketStore, int minBucketRedistributionIntervalInSeconds,
+                               BucketDistributor bucketDistributor, BucketStore.ServiceType serviceType,
+                               ScheduledExecutorService executorService) {
+        Preconditions.checkNotNull(bucketStore, "bucketStore");
+        Preconditions.checkArgument(minBucketRedistributionIntervalInSeconds >= 0,
+                "minBucketRedistributionInterval should not be negative");
+
+        this.bucketStore = bucketStore;
+        this.minBucketRedistributionIntervalInSeconds = Duration.ofSeconds(minBucketRedistributionIntervalInSeconds);
+        this.bucketDistributor = bucketDistributor;
+        this.serviceType = serviceType;
+        this.executorService = executorService;
+    }
+
+    /**
+     * Suspend the leader thread.
+     */
+    public void suspend() {
+        suspended.set(true);
+    }
+
+    /**
+     * Resume the suspended leader thread.
+     */
+    public void resume() {
+        if (suspended.compareAndSet(true, false)) {
+            suspendMonitor.release();
+        }
+    }
+
+    @Override
+    public void takeLeadership(CuratorFramework client) throws Exception {
+        log.info("{}: Obtained leadership to monitor the controller to buckets mapping.", serviceType);
+
+        // Attempt a distribution whenever leadership is obtained to ensure no Controllers events are missed.
+        controllerChange.release();
+
+        // Start cluster monitor.
+        synchronized (lock) {
+            pravegaServiceCluster = new ClusterZKImpl(client, ClusterType.CONTROLLER);
+        }
+
+        // Add listener to track controller changes on the monitored pravega cluster.
+        pravegaServiceCluster.addListener((type, controller) -> {
+            switch (type) {
+                case HOST_ADDED:
+                case HOST_REMOVED:
+                    // We don't keep track of the controllers and we always query for the entire set from the cluster
+                    // when changes occur. This is to avoid any inconsistencies if we miss any notifications.
+                    log.info("{}: Received event: {} for host: {}. Wake up leader for distribution", serviceType, type, controller);
+                    controllerChange.release();
```

Comment:
This may be releases spuriously due to a past listener left over from a previous invocation of `takeLeadership`. I am not clear how that is being handled.

Reply:
When we receive events about changes in pravega cluster(ADDED/REMOVED) then we always query for entire set of controller in **triggerDistribution()** method.
```
 synchronized (lock) {
            currentControllers = pravegaServiceCluster.getClusterMembers().stream().map(controller ->
                    controller.getHostId()).collect(Collectors.toSet());
        }
```
This will ensure us there is no inconsistencies and distribution happens as per latest state of pravega cluster.